File: MaWPictures/src/main/java/us/mikeandwan/photos/services/PhotoApi.java
package us.mikeandwan.photos.services;

import java.util.List;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import us.mikeandwan.photos.models.Category;
import us.mikeandwan.photos.models.Comment;
import us.mikeandwan.photos.models.CommentPhoto;
import us.mikeandwan.photos.models.ExifData;
import us.mikeandwan.photos.models.Photo;
import us.mikeandwan.photos.models.PhotoAndCategory;
import us.mikeandwan.photos.models.RatePhoto;
import us.mikeandwan.photos.models.Rating;


interface PhotoApi {
    /*
    @GET("photos/getPhotoYears")
    Call<List<Integer>> getPhotoYears();

    @GET("photos/getCategoriesForYear/{year}")
    Call<List<Category>> getCategoriesForYear(@Path("year") int year);
    */

    @GET("photos/getRecentCategories/{sinceId}")
    Call<List<Category>> getRecentCategories(@Path("sinceId") int sinceId);

    /*
    @GET("photos/getCategoryCount")
    Call<Integer> getTotalCategoryCount();
    */

    @GET("photos/getPhotoExifData/{photoId}")
    Call<ExifData> getExifData(@Path("photoId") int photoId);

    @GET("photos/getRandomPhoto")
    Call<PhotoAndCategory> getRandomPhoto();

    @GET("photos/getCommentsForPhoto/{photoId}")
    Call<List<Comment>> getComments(@Path("photoId") int photoId);

    @GET("photos/getRatingForPhoto/{photoId}")
    Call<Rating> getRatings(@Path("photoId") int photoId);

    @GET("photos/getPhotosByCategory/{categoryId}")
    Call<List<Photo>> getPhotosByCategory(@Path("categoryId") int categoryId);

    @GET("photos/getPhotosByCommentDate/{newestFirst}")
    Call<List<Photo>> getPhotosByCommentDate(@Path("newestFirst") boolean newestFirst);

    @GET("photos/getPhotosByUserCommentDate/{newestFirst}")
    Call<List<Photo>> getPhotosByUserCommentDate(@Path("newestFirst") boolean newestFirst);

    @GET("photos/getPhotosByCommentCount/{mostFirst}")
    Call<List<Photo>> getPhotosByCommentCount(@Path("mostFirst") boolean mostFirst);

    @GET("photos/getPhotosByAverageRating/true")
    Call<List<Photo>> getPhotosByAverageRating();

    @GET("photos/getPhotosByUserRating/true")
    Call<List<Photo>> getPhotosByUserRating();

    @POST("photos/ratePhoto")
    Call<Float> ratePhoto(@Body RatePhoto rating);

    @POST("photos/addCommentForPhoto")
    Call<Boolean> addCommentForPhoto(@Body CommentPhoto commentPhoto);
}

File: MaWPhotos/src/main/kotlin/us/mikeandwan/photos/api/PhotoApi.kt
package us.mikeandwan.photos.api

import retrofit2.Response
import retrofit2.http.*

internal interface PhotoApi {
    @GET("photo-categories/recent/{sinceId}")
    suspend fun getRecentCategories(@Path("sinceId") sinceId: Int): Response<ApiCollection<Category>>

    @GET("photos/{photoId}/exif")
    suspend fun getExifData(@Path("photoId") photoId: Int): Response<ExifData>

    @GET("photos/random/{count}")
    suspend fun getRandomPhotos(@Path("count") count: Int): Response<ApiCollection<Photo>>

    @GET("photos/{photoId}/comments")
    suspend fun getComments(@Path("photoId") photoId: Int): Response<ApiCollection<Comment>>

    @GET("photos/{photoId}/rating")
    suspend fun getRatings(@Path("photoId") photoId: Int): Response<Rating>

    @GET("photo-categories/{categoryId}/photos")
    suspend fun getPhotosByCategory(@Path("categoryId") categoryId: Int): Response<ApiCollection<Photo>>

    @PATCH("photos/{photoId}/rating")
    suspend fun ratePhoto(@Path("photoId") photoId: Int, @Body rating: RatePhoto): Response<Rating>

    @POST("photos/{photoId}/comments")
    suspend fun addCommentForPhoto(
        @Path("photoId") photoId: Int,
        @Body commentPhoto: CommentPhoto
    ): Response<ApiCollection<Comment>>
}
